Hour Dashboard Component

vinitbytheminute opened the pull request:
### ### Hour Widget 

**Purpose**
The hour widget will track the hours and nights of each member in the group. The widget allows a member to toggle between the hour and total view to see the number of hours spent and nights slept in Kville for that week or in total for tenting. There also is a dropdown that allows a user to sort the members by Hours or Night within a particular view.
 
Currently, the widget displays mock data and does not feed actual hours and nights of the tent members into the display. This PR is only front-end focused, and the backend will be done later on.

**Testing**
1. Sign in with an account and navigate to the dashboard
2.  Change View from Week to Total to see the scripts of each card change and order to the rank already selected
3. Toggle between different Ranks in order to sort the member cards in a particular order 

**Screenshots (If Needed)**
<img width="1259" alt="screen shot 2018-12-25 at 12 09 06 pm" src="https://user-images.githubusercontent.com/44485851/50427791-0e87a300-0865-11e9-8381-2b1b9e96d892.png">

rikkikendall committed:
working on FAQ

rikkikendall committed:
stillrouting

rikkikendall committed:
update

rikkikendall committed:
Update Tenting101

rikkikendall committed:
still working

rikkikendall committed:
completedrouting

rikkikendall committed:
updatetenting101

rikkikendall committed:
completed

rikkikendall committed:
Merge branch 'master' of https://github.com/GTHC/web into FAQ-rikkikendall

rikkikendall committed:
what up

amanmibra committed:
Updated gitignore

amanmibra committed:
Made styling adjustments

amanmibra committed:
Merge branch 'master' of github.com:GTHC/web into FAQ-rikkikendall

amanmibra committed:
Fix error messages that appeared after rendering static files

vinitbytheminute committed:
Fixing up the sorting

vinitbytheminute committed:
Got rid of Hour Component. Will add this once WIP:DB_H_W PR addressed

mafuvadze commented on app/javascript/src/components/Dashboard/Hours.jsx:
```diff
@@ -0,0 +1,326 @@
+      rank = view == "hour" ? user.rank.whour : user.rank.wnight;
+    }
+    if (rank == 1){
+      classes+= "warning";
```
Not a fan of using these defaults. I think you should just define your own colors and only use warning attributes for warnings.

mafuvadze commented on app/javascript/src/components/Dashboard/Hours.jsx:
```diff
@@ -0,0 +1,326 @@
+// Semantic Ui Components Used
+import { Dropdown, Card, Grid, Button, Header, Icon} from "semantic-ui-react";
+//Endpoints not yet created, so Created Fake Data Instead
+import {getHours} from "./fakeData";
```
Classes are typically not snake case. Change fakeData -> MockData

mafuvadze commented on app/javascript/src/components/Dashboard/Hours.jsx:
```diff
@@ -0,0 +1,326 @@
+  constructor(props) {
+  super(props);
+
+  this.state = {
```
state attributes need to in camel case. Hours -> hours, Rank -> rank, etc

mafuvadze commented on app/javascript/src/components/Dashboard/Hours.jsx:
```diff
@@ -0,0 +1,326 @@
+
+updateTotal = (Total) => {
+  this.forceUpdate();
+  console.log(this.state.Total);
```
remove this. Bad coding practice to leave console.logs in prod code

mafuvadze commented on app/javascript/src/components/Dashboard/Hours.jsx:
```diff
@@ -0,0 +1,326 @@
+
+
+//Function sorting Data from Greatest to Least by the given Rank and Specific View (Week or Total)
+handleNightorHours = (e, timeOfDay) => {
```
what is e? make this arg more descriptive

mafuvadze commented on app/javascript/src/components/Dashboard/Hours.jsx:
```diff
@@ -0,0 +1,326 @@
+
+//Function sorting Data from Greatest to Least by the given Rank and Specific View (Week or Total)
+handleNightorHours = (e, timeOfDay) => {
+  console.log(timeOfDay);
```
remove

mafuvadze commented on app/javascript/src/components/Dashboard/Hours.jsx:
```diff
@@ -0,0 +1,326 @@
+//Function sorting Data from Greatest to Least by the given Rank and Specific View (Week or Total)
+handleNightorHours = (e, timeOfDay) => {
+  console.log(timeOfDay);
+  console.log(this.state.Total);
```
remove

mafuvadze commented on app/javascript/src/components/Dashboard/Hours.jsx:
```diff
@@ -0,0 +1,326 @@
+      return timeB.week - timeA.week;
+    });
+  }
+  console.log(newNight)
```
remove

mafuvadze commented on app/javascript/src/components/Dashboard/Hours.jsx:
```diff
@@ -0,0 +1,326 @@
+
+//Each User Hours and Night count displayed for a given view
+  CustomCard = (props) => {
+    console.log(this.state.Total);
```
remove

mafuvadze commented on app/javascript/src/components/Dashboard/Hours.jsx:
```diff
@@ -0,0 +1,326 @@
+    ]
+    const { Total, Hours } = this.state;
+    return (
+    <Card fluid >
```
Really long. Consider breaking it into components and importing them

vinitbytheminute replied:
Quite difficult to break up into components. Please look at new code to see if this is still an issue

mafuvadze commented on app/javascript/src/components/Dashboard/fakeData.js:
```diff
@@ -0,0 +1,79 @@
+  {
+    name: "James ",
+    hours: {week: 17, total: 132},
+    Night: {week: 1, total: 13},
```
irregular naming scheme. Stick to camel case

mafuvadze reviewed (changes requested):
CODE:
Left some inline comments. Overall I would work on the naming scheme. Make variables and functions more descriptive and stick to camelCase. The code may also benefit from being split into multiple components or functions.

UI:
- I had made the upcoming shifts using a slightly different UI pattern (with the title on the left). Can you make it like that so we have consistency.
- The drop-downs look weird centered and on top of each other. They should be on the left and side by side. 
- Color ranking isn't working

vinitbytheminute committed:
Merge branch 'staging' of https://github.com/GTHC/web into dashboard-hours

vinitbytheminute committed:
New push

vinitbytheminute committed:
Cleaning Up Code ( moved Dashboard into seperate function )

rikkikendall committed:
Updated navbar on tenting101

rikkikendall committed:
Update About.jsx

rikkikendall committed:
Merge branch 'FAQ-rikkikendall' of https://github.com/GTHC/web into FAQ-rikkikendall

rikkikendall committed:
Merge branch 'FAQ-rikkikendall' of https://github.com/GTHC/web into FAQ-rikkikendall

amanmibra committed:
Built API endpoint to find user aprox availability for a new shift

amanmibra committed:
Translating JS date objects to availability indicies

amanmibra committed:
Added availability feature to CreateShiftForm

amanmibra committed:
Moved getShiftAvail util

amanmibra committed:
Added availability feature to UpdateShiftForm

vinitbytheminute committed:
Changed labels to tophies, Dropdown now inline, altered rank script

amanmibra committed:
Added db migrate script

vinitbytheminute committed:
UI Consistent Across MyShifts and Hours Component on dashboard now

amanmibra commented:
I am in the middle of reviewing this, and I notice a simple compilation error, which I will fix myself, but you put the Hours component in a folder called "Dashboard", but called from "dashboard". Please, be careful, imports are case-sensitive.  @vinitbytheminute

amanmibra committed:
Removed "Dashboard" folder and moved contents to "dashboard"

amanmibra reviewed (changes requested):
I don't think I am ready to even do a technical review on this as the design of the widget is a bit confusing. We should discuss this as a team sometime this week.

Looking at this, tbh, I do not understand want any of these numbers mean:

![image](https://user-images.githubusercontent.com/13587407/50427254-f6b61c00-0870-11e9-968c-1a1222d09df5.png)

In this example, what is the difference between "Total Nights" and "Nights Spent", and why is "Total Nights" less than "Nights Spent"? When building out UI, you need to accommodate for even the "slowest" of people, so, when they look at it they don't need explanation. I had to look at this for awhile before I kindof had an idea of what's going on.

### Most Importantly!
When you are pushing to `staging`, you are essentially saying what I have built is ready for production! But, you are using mock data (understandably so) in this PR, and that is a huuuuuge no-no for production. Like I mentioned when I had you do the you `availability-vinit` PR, when you are building front end components that are NOT connected to backend, we will have you push into a branch such as `availability-amanmibra`, so, I, or hopefully eventually anyone else working backend, can connect your work to the database. I went ahead, and made `dashboard-hours-amanmibra`, and I will change your base branch to it.

rikkikendall committed:
routed

rikkikendall committed:
step is functional

amanmibra committed:
Make Sign Up Stepper responsive (unrelated fix)

amanmibra committed:
Added session check to user_controller

amanmibra committed:
Added session check to app rendering

vinitbytheminute commented:
@amanmibra  That was a wonderful point you brought up regarding the confusion of Total Nights and Nights Slept. During the User testing, A user suggested I switch Night Rank to Nights Slept as he found it intuitive that the bottom section meant ranking. However, after reading your comment, it is clear that the bottom section of the card does not indicate to the user that it means rank without explicitly stating that. Therefore I changed the UI to this:

<img width="536" alt="screen shot 2018-12-26 at 4 20 23 pm" src="https://user-images.githubusercontent.com/44485851/50460351-37be3700-092a-11e9-9ff2-625a64cf4627.png">

Another View I suggest is this. It shows What view the user is on (Total or Week) within the card. 
<img width="517" alt="screen shot 2018-12-26 at 4 37 50 pm" src="https://user-images.githubusercontent.com/44485851/50460635-b1572480-092c-11e9-92f8-d89a71011257.png">

vinitbytheminute committed:
Added Suttle Change to UI to clarify ranking section of member card

amanmibra commented:
@vinitbytheminute That looks much better, and easier to read from the screenshot. Tomorrow morning I will do a full review.

rikkikendall committed:
Merge branch 'staging' of https://github.com/GTHC/web into FAQ-rikkikendall

amanmibra committed:
Made the following changes to Availability:
- Key Redesign
- Changed location of time rows
- Removed lodash
- Updated API requests

amanmibra committed:
Added popup for color info about user dropdown

amanmibra committed:
Added passcode to TeamProfile info

amanmibra committed:
Remove console.logs

rikkikendall committed:
DB updated, almost complete

rikkikendall committed:
addedaboutpage

rikkikendall committed:
revised

rikkikendall committed:
organizedfiles

rikkikendall committed:
componentsorganized

rikkikendall committed:
organizedcodebase

rikkikendall committed:
allfinished

rikkikendall committed:
updatedcss

amanmibra reviewed (dismissed):
Alright looks good to go to `dashboard-hours-amanmibra`!

amanmibra committed:
Merge branch 'staging' of github.com:GTHC/web into dashboard-hours

amanmibra commented on app/javascript/src/components/dashboard/Hours.jsx:
```diff
@@ -0,0 +1,242 @@
+  }
+
+  //Needed to updated 'total' state value while switching view in dropdown. Look at line 192
+  handleChange(e, timeofDay) {
```
Arrow functions please. We are 100% ES6 only. Also `timeofDay` is not camel cased. Work on that please.

```
handleChange = (e, timeOfDay) => {
...
}
```

amanmibra commented on app/javascript/src/components/dashboard/Hours.jsx:
```diff
@@ -0,0 +1,242 @@
+      : user.rank.thour;
+  };
+
+  nightRank(user) {
```
Arrow functions please

amanmibra commented on app/javascript/src/components/dashboard/Hours.jsx:
```diff
@@ -0,0 +1,242 @@
+  };
+
+  //Function determining which Rank Data to Display
+  hourRank(user) {
```
Arrow func

amanmibra commented on app/javascript/src/components/dashboard/Hours.jsx:
```diff
@@ -0,0 +1,242 @@
+      : user.hours.total;
+  };
+
+  nightNumber(user) {
```
same thing, arrow functions

amanmibra replied:
Take the time to find the rest of your functions that are not arrow functions. I see at least two others that I didn't point out.

amanmibra commented on app/javascript/src/components/dashboard/Hours.jsx:
```diff
@@ -0,0 +1,242 @@
+  }
+
+  //begining of render function
+  render() {
```
Before you are confused, all of your functions except the extended React library functions (`render()`, `componentWillMount()`, `constructor()`, and pretty much all of the functions on [the react docs here](https://reactjs.org/docs/react-component.html)), should be arrow functions. So, this line is fine and does not need to be an arrow function.

amanmibra commented on app/javascript/src/components/dashboard/Hours.jsx:
```diff
@@ -0,0 +1,242 @@
+
+  // Trying to move the dropDown options out of the render component, but dropDown fucntion  is buggy
+
+  dropDown = () => {
```
I would consider making this and the individual Cards' content left aligned. Just for consistency.

amanmibra commented on app/javascript/src/components/dashboard/MyShifts.jsx:
```diff
@@ -1,7 +1,7 @@
 import React, { Component } from 'react';
```
Please remove all changes you made to this component. Your changes have made some inconsistencies in styling.

amanmibra reviewed (changes requested):
Yea, there are some remaining changes that you need to make. If you don't have time to do it by tonight, just at least look at my requests so you can learn from your mistakes, and then lmk so I can fix it, it will only take me a few minutes. 

Also, please look at your console when building components, they give you warnings whenever you are using a component or its props incorrectly, and you have a ton of warnings/errors in your console from this component.

amanmibra commented on app/javascript/src/components/dashboard/Hours.jsx:
```diff
@@ -0,0 +1,242 @@
+  }
+
+  //begining of render function
+  render() {
```
Please change the render function to the following:

```
render() {
    const { hours } = this.state;
    return (
      <Card fluid>
        <Card.Content>
          <Card.Header>Hour Breakdown</Card.Header>
        </Card.Content>
        <Card.Content>
          {this.dropDown()}
          <br/>
          <br/>
          <Card.Group>
            {  
              hours.map((user) => (this.customCard({ user, key: user.name })))
            }
          </Card.Group>
        </Card.Content>
      </Card>
    );
  }
}
```
I had typed a list of reasons why earlier, but I did not save my previous comment, so, just ask me in the meeting tonight and I will tell you. @vinitbytheminute

amanmibra replied:
Actually, I will type it out:
- You used class functions as components, `<this.dropDown />` and `<this.customCard />`. You cannot do that. In classes, functions stay functions, if you want to build a separate component, then build it outside of the class.
- Use Card.Header, it messes with the styling when you do otherwise
- You called a variable from the state, that you didn't even use. 
 
This is not everything I typed earlier, but just ask me when you get the chance and there are some conventions that I need to go over with you.

amanmibra reviewed (changes requested):
Yea, there are some remaining changes that you need to make. If you don't have time to do it by tonight, just at least look at my requests so you can learn from your mistakes, and then lmk so I can fix it, it will only take me a few minutes. 

Also, please look at your console when building components, they give you warnings whenever you are using a component or its props incorrectly, and you have a ton of warnings/errors in your console from this component.

amanmibra committed:
Update the sess check action's isLoading element to go false after request

amanmibra committed:
Merge pull request #40 from GTHC/availability-sort-amanmibra

Availability Labels to Shift Users

amanmibra committed:
Merge branch 'staging' of github.com:GTHC/web into iss-41-amanmibra

vinitbytheminute committed:
Made Adjustment that correspond to Aman's suggestions on PR 37.

vinitbytheminute committed:
Removed unrecognized string values from boolean props in dropdown

rikkikendall committed:
fixed-PR-comments

amanmibra committed:
Updated ShiftViewModal to improve data visualization

amanmibra committed:
Updated heroku.yml

amanmibra committed:
Setting up for heroku build and release

amanmibra committed:
Merge pull request #42 from GTHC/iss-41-amanmibra

Fix "Rails Auth not carried over to Redux"

amanmibra committed:
Merge branch 'staging' of github.com:GTHC/web into update-shift-views

amanmibra committed:
Change sign up email label to "Duke Email"

amanmibra committed:
Change date string formatting

amanmibra committed:
Created ShiftTimeInput and implemented to creating shifts

amanmibra committed:
Moved html breaks

rikkikendall committed:
NavBar-Changed-White-Login

amanmibra committed:
Merge branch 'staging' of github.com:GTHC/web into FAQ-rikkikendall

amanmibra committed:
Merge pull request #43 from GTHC/FAQ-rikkikendall

Updated Tenting101

rikkikendall committed:
Merge branch 'staging' of https://github.com/GTHC/web into About-rikkikendall

amanmibra committed:
Added ShiftTimeInput onto update shifts components

amanmibra committed:
Improved spacing between time and note descriptions in View Modal

amanmibra commented:
Request a review when you think its ready @vinitbytheminute

amanmibra committed:
Merge branch 'staging' into update-shift-views

amanmibra committed:
Merge pull request #44 from GTHC/update-shift-views

Add details to each shift modals

rikkikendall committed:
about-page-content-finished

rikkikendall committed:
about-page-routing-complete

Completed routing for the about page, added routing on tenting101 to the about page, and edited the about page css.

rikkikendall committed:
Merge branch 'staging' of https://github.com/GTHC/web into About-rikkikendall

amanmibra commented:
Yo it's been 10 days @vinitbytheminute

amanmibra committed:
Added to .gitignore and fixed some console warnings

amanmibra committed:
Merge pull request #45 from GTHC/Tenting-rikkikendall

About Page

amanmibra committed:
Removed user_id from schema.rb

amanmibra commented on app/javascript/src/components/dashboard/Hours.jsx:
```diff
@@ -0,0 +1,233 @@
+
+  //Function created to Switch between Total and Week View script for both Hour and Night rank
+
+  hourView() {
```
arrow functions @vinitbytheminute

amanmibra commented on app/javascript/src/components/dashboard/Hours.jsx:
```diff
@@ -0,0 +1,233 @@
+      : "Hours Spent: ";
+  };
+
+  nightView() {
```
again, arrow functions

amanmibra commented on app/javascript/src/components/dashboard/Hours.jsx:
```diff
@@ -0,0 +1,233 @@
+  }
+
+  //Needed to updated 'total' state value while switching view in dropdown. Look at line 192
+  handleChange(e, timeOfDay) {
```
arrow functions

amanmibra reviewed (commented):
@vinitbytheminute 
You forgot some arrow functions...

I will just change it myself and merge.

amanmibra committed:
Added arrow functions

amanmibra committed:
Merge branch 'staging' of github.com:GTHC/web into dashboard-hours

amanmibra commented:
Again, always remember `git pull origin staging`
